### PR TITLE
Clarification to the documentation on converting payable to non payable functions.

### PR DESCRIPTION
**Pull request by Andy53**
Added clarification to the documentation as described in [#13114](https://github.com/ethereum/solidity/issues/13114)

### CODE REVIEW

File: docs/types/value-types.rst [hunk 1]
```diff
@@ -770,6 +770,16 @@ confusing, but in essence, if a function is ``payable``, this means that it
...
+To clarify not rejecting ether is less restrictive than rejecting ether
+therefore you can override a payable function with non-payable but not the
+other way around
+
+Additionally When you define a ``non-payable`` function pointer, the compiler
+does not enforce that the function will actually reject ether. Instead it
+enforces that the function pointer is never used to send ether. Therefore
+it is possible to assign a ``payable`` function pointer to a ``non-payable``
+pointer due to the fact that you can't send ether ensuring both types behave
+the same way.
```
Comment (NunoFilipeSantos) [inline docs/types/value-types.rst]:
```suggestion
To clarify, not rejecting ether is less restrictive than rejecting ether.
This means you can override a payable function with a non-payable but not the
other way around

Additionally, when you define a ``non-payable`` function pointer, 
the compiler does not enforce that the function will actually reject ether. 
Instead, it enforces that the function pointer is never used to send ether. 
Which makes it possible to assign a ``payable`` function pointer to a ``non-payable``
pointer due to not being able to send ether ensuring both types behave the same way.
```

**Review by NunoFilipeSantos — CHANGES_REQUESTED** [review]
@christianparpart take a look here.

File: docs/types/value-types.rst (hunk 1 above)
Comment (christianparpart) [reply docs/types/value-types.rst]: :+1:

File: docs/types/value-types.rst [hunk 2]
```diff
@@ -770,6 +770,15 @@ confusing, but in essence, if a function is ``payable``, this means that it
 also accepts a payment of zero Ether, so it also is ``non-payable``.
 On the other hand, a ``non-payable`` function will reject Ether sent to it,
 so ``non-payable`` functions cannot be converted to ``payable`` functions.
+To clarify, not rejecting ether is less restrictive than rejecting ether.
+This means you can override a payable function with a non-payable but not the
+other way around
```
Comment (matheusaaguiar) [inline docs/types/value-types.rst]:
```suggestion
other way around.
```

File: docs/types/value-types.rst [hunk 3]
```diff
@@ -770,6 +770,15 @@ confusing, but in essence, if a function is ``payable``, this means that it
 also accepts a payment of zero Ether, so it also is ``non-payable``.
 On the other hand, a ``non-payable`` function will reject Ether sent to it,
 so ``non-payable`` functions cannot be converted to ``payable`` functions.
+To clarify, not rejecting ether is less restrictive than rejecting ether.
```
Comment (matheusaaguiar) [inline docs/types/value-types.rst]:
```suggestion
To clarify, rejecting ether is more restrictive than not rejecting ether.
```
I feel that this is more straightforward.

File: docs/types/value-types.rst [hunk 4]
```diff
@@ -770,6 +770,15 @@ confusing, but in essence, if a function is ``payable``, this means that it
...
+This means you can override a payable function with a non-payable but not the
+other way around
+
+Additionally, when you define a ``non-payable`` function pointer, 
+the compiler does not enforce that the function will actually reject ether.
```
Comment (matheusaaguiar) [inline docs/types/value-types.rst]:
```suggestion
the compiler does not enforce that the pointed function will actually reject ether. 
```

File: docs/types/value-types.rst [hunk 5]
```diff
@@ -770,6 +770,15 @@ confusing, but in essence, if a function is ``payable``, this means that it
...
 so ``non-payable`` functions cannot be converted to ``payable`` functions.
+To clarify, not rejecting ether is less restrictive than rejecting ether.
+This means you can override a payable function with a non-payable but not the
+other way around
+
+Additionally, when you define a ``non-payable`` function pointer, 
+the compiler does not enforce that the function will actually reject ether. 
+Instead, it enforces that the function pointer is never used to send ether. 
+Which makes it possible to assign a ``payable`` function pointer to a ``non-payable``
+pointer due to not being able to send ether ensuring both types behave the same way.
```
Comment (matheusaaguiar) [inline docs/types/value-types.rst]:
```suggestion
function pointer ensuring both types behave the same way, i.e, both cannot be used 
to send ether.
```

**Review by matheusaaguiar — COMMENTED** [review]
I guess this is confusing because it involves two different situations: functions and pointers to functions. I think that the difference is more apparent now in the text, but maybe there is room to be even more explicit.

**Comment by matheusaaguiar** [comment]
Hi @Andy53 , it seems that in your last commit you mistakenly merged branch `develop` on your PR branch instead of rebasing on top of the former.
You could fix it by first rewinding the last commit and then rebasing on top of develop.
```
git reset --merge HEAD~1
git rebase -i origin/develop
```
The rebase command with `-i` should open an interactive session in which you can choose to squash some commits, like the `trailing whitespace` ones, for example.
Feel free to ask or dm if you need help.

**Comment by Andy53** [comment]
This rebase did not go to plan sadly :( I seem to have pulled 2 commits in from another branch.

**Comment by Andy53** [comment]
So the latest push [4357949](https://github.com/ethereum/solidity/commit/43579491e3ba267d23afaaefdd768e7906bd7973) seems to be correct. It's got just the changes to the documentation required for the PR and nothing else.

**Comment by matheusaaguiar** [comment]
Sorry, maybe I misdirect you with the commands I suggested. Probably, the `origin/develop` you rebased onto is actually your own fork repository. Maybe you have a `upstream` remote which points to the solidity original repo? If not, you can add it with:
```
git remote add upstream https://github.com/ethereum/solidity.git
git fetch upstream
```
Anyway, there are still commits not belonging and also the last one you mentioned, [4357949](https://github.com/ethereum/solidity/commit/43579491e3ba267d23afaaefdd768e7906bd7973), is your first modification.
Try to see if you can still find the other relevant commits and then rebase on the `develop` branch of the original solidity repo. Otherwise, I have resumed the changes so far in my local copy and can push them later.

**Comment by Andy53** [comment]
I think I might have got it right this time. Files changed is down to 1 now.

**Comment by matheusaaguiar** [comment]
Restored the lost modifications and rebased on top of `develop`.

**Review by NunoFilipeSantos — DISMISSED** [review]
Thank you.

**Review by matheusaaguiar — APPROVED** [review]
I have rebased and squashed commits. Setting to auto-merge, since @NunoFilipeSantos  had previously approved.

**Pull request merged**